midend/lib/Utils/DIPUtils.cpp

```diff
@@ -410,6 +411,185 @@ Value customTanVal(OpBuilder &builder, Location loc, Value angleVal) {
+      });
+  D = scfRes.getResult(0);
+  Value negD = builder.create<arith::NegFOp>(loc, D);
+  Value A11 = builder.create<arith::MulFOp>(loc, affineMatrix[4], D);
+  Value A22 = builder.create<arith::MulFOp>(loc, affineMatrix[0], D);
+  affineMatrix[0] = A11;
+  affineMatrix[1] = builder.create<arith::MulFOp>(loc, affineMatrix[1], negD);
+  affineMatrix[3] = builder.create<arith::MulFOp>(loc, affineMatrix[3], negD);
+  affineMatrix[4] = A22;
+  Value m0pm2 =
+      builder.create<arith::MulFOp>(loc, affineMatrix[0], affineMatrix[2]);
+  Value m1pm5 =
+      builder.create<arith::MulFOp>(loc, affineMatrix[1], affineMatrix[5]);
+  Value negB1 = builder.create<arith::AddFOp>(loc, m0pm2, m1pm5);
+  Value m2pm3 =
+      builder.create<arith::MulFOp>(loc, affineMatrix[2], affineMatrix[3]);
+  Value m4pm5 =
+      builder.create<arith::MulFOp>(loc, affineMatrix[4], affineMatrix[5]);
+  Value negB2 = builder.create<arith::AddFOp>(loc, m2pm3, m4pm5);
+  affineMatrix[2] = builder.create<arith::NegFOp>(loc, negB1);
+  affineMatrix[5] = builder.create<arith::NegFOp>(loc, negB2);
+}
+
+void affineTransformController(OpBuilder &builder, Location loc,
+                               MLIRContext *ctx, Value input, Value output,
+                               SmallVector<Value, 6> affineMatrix,
+                               int64_t stride) {
+  VectorType vectorTyF32 = VectorType::get({stride}, FloatType::getF32(ctx));
+
+  Value c0 = builder.create<arith::ConstantIndexOp>(loc, 0);
+  Value c1 = builder.create<arith::ConstantIndexOp>(loc, 1);
+  Value c512F32 =
+      builder.create<arith::ConstantOp>(loc, builder.getF32FloatAttr(512));
+  Value c1024F32 =
+      builder.create<arith::ConstantOp>(loc, builder.getF32FloatAttr(1024));
+  Value c512F32Vec = builder.create<vector::SplatOp>(loc, vectorTyF32, c512F32);
+  Value c1024F32Vec =
+      builder.create<vector::SplatOp>(loc, vectorTyF32, c1024F32);
+
+  inverseAffineMatrix(builder, loc, affineMatrix);
+
+  Value m0Vec =
+      builder.create<vector::SplatOp>(loc, vectorTyF32, affineMatrix[0]);
+  Value m1Vec =
+      builder.create<vector::SplatOp>(loc, vectorTyF32, affineMatrix[1]);
+  Value m2Vec =
+      builder.create<vector::SplatOp>(loc, vectorTyF32, affineMatrix[2]);
+  Value m3Vec =
+      builder.create<vector::SplatOp>(loc, vectorTyF32, affineMatrix[3]);
+  Value m4Vec =
+      builder.create<vector::SplatOp>(loc, vectorTyF32, affineMatrix[4]);
+  Value m5Vec =
+      builder.create<vector::SplatOp>(loc, vectorTyF32, affineMatrix[5]);
+
+  Value inputRow = builder.create<memref::DimOp>(loc, input, c0);
+  Value inputCol = builder.create<memref::DimOp>(loc, input, c1);
+
+  Value outputRow = builder.create<memref::DimOp>(loc, output, c0);
+  Value outputCol = builder.create<memref::DimOp>(loc, output, c1);
+
+  Value strideVal = builder.create<arith::ConstantIndexOp>(loc, stride);
+
+  // generate vector [0., 1., ..., stride - 1]
+  std::vector<float> xInit(stride);
```

Comment:
Can you consider adding a generic function for this in `Utils.cpp`? ([Ref](https://github.com/buddy-compiler/buddy-mlir/blob/cc57e60f073d769224a334b4df10c16b43eb2711/midend/lib/Utils/Utils.cpp#L102))

Reply:
this method cannot be used if the start index is a Value. I initialize it using std::vector<float>, not std::vecctor<Value>

Reply:
Okay. I'll add this in Utils.cpp.